ipn/ipnlocal/profiles.go

```diff
@@ -304,31 +351,78 @@ func (pm *profileManager) writePrefsToStore(key ipn.StateKey, prefs ipn.PrefsVie
 	return nil
 }
 
-// Profiles returns the list of known profiles.
+// Profiles returns the list of known profiles accessible to the current user.
 func (pm *profileManager) Profiles() []ipn.LoginProfile {
 	allProfiles := pm.allProfiles()
-	out := make([]ipn.LoginProfile, 0, len(allProfiles))
-	for _, p := range allProfiles {
-		out = append(out, *p)
+	out := make([]ipn.LoginProfile, len(allProfiles))
+	for i, p := range allProfiles {
+		out[i] = *p
 	}
 	return out
 }
 
+// ProfileByID returns a profile with the given id, if it is accessible to the current user.
+// If the profile exists but is not accessible to the current user, it returns an [errProfileAccessDenied].
+// If the profile does not exist, it returns an [errProfileNotFound].
+func (pm *profileManager) ProfileByID(id ipn.ProfileID) (ipn.LoginProfile, error) {
+	kp, err := pm.profileByIDNoCheck(id)
+	if err != nil {
+		return ipn.LoginProfile{}, err
+	}
+	if err := pm.checkProfileAccess(kp); err != nil {
+		return ipn.LoginProfile{}, err
+	}
+	return *kp, nil
+}
+
+func (pm *profileManager) profileByIDNoCheck(id ipn.ProfileID) (*ipn.LoginProfile, error) {
+	if id == pm.currentProfile.ID {
+		return pm.currentProfile, nil
+	}
+	kp, ok := pm.knownProfiles[id]
+	if !ok {
+		return nil, errProfileNotFound
+	}
+	return kp, nil
+}
+
+// ProfilePrefs returns preferences for a profile with the given id.
+// If the profile exists but is not accessible to the current user, it returns an [errProfileAccessDenied].
```

Comment:
do we care about leaking the existence of a profile here?

Reply:
I don't think we do. It's just a random 4-digit identifier. Even if we returned this error to the user and they enumerated all possible IDs, the only information leaked would be that another user has a Tailscale profile on a shared device with Tailscale installed. And we already expose this, as we prevent other users from connecting when the LocalAPI is in use by a different Windows user.

But we'll refine it with the new permission model to return different errors depending on whether the caller has read access to the profile.